keps/sig-cluster-lifecycle/kubeadm/1739-customization-with-patches/README.md

```diff
@@ -0,0 +1,626 @@
+-->
+
+### Code organization
+
+The code organization of the new feature will be done in a similar
+fashion to the existing Kustomize feature. The backend will be stored
+in a new package called `customize`. Kubeadm commands would have minimal
+exposure to the underlying backend and only call a single method that
+accepts the directory where the user patches are.
+
+### Patch formats
+
+The proposed design is supporting the following patch formats similarly
+to kubectl:
+- Strategic Merge Patches (default):
+  - Supported as `strategic` by `kubectl patch`.
+  - Will be implemented using the `k8s.io/apimachinery` library.
+- RFC6902 JSON patches:
+  - Supported as `json` by `kubectl patch`.
+  - Would be implemented using the `github.com/evanphx/json-patch`
+  library.
+- RFC7396 JSON merge patches:
+  - Supported as `merge` by `kubectl patch`.
+  - Would be implemented using the `github.com/evanphx/json-patch`
+  library.
+
+Note that kubectl already uses the same backends for the above
+patch formats.
+
+Patches defined in both YAML and JSON will be supported, with
+the exception of `json` which must be in JSON only.
+
+Conversion from YAML to JSON will be performed using the
+`sigs.k8s.io/yaml` library.
+
+### Patch file naming format
+
+Once the user passes a folder with patches to a kubeadm command,
+kubeadm must be able to determine what format each file is and what
+manifest it is targeting.
+
+Parsing the patch contents to determine the patch format is not possible,
+because the formats `strategic` and `merge` are the exactly the same,
+while they do have different apply mechanisms.
+
+The proposal is to have the following naming format:
+`componentname[suffix][+patchtype][.extension]`
+
+- `componentname` must be a known component name - e.g. `kube-apiserver`.
+- `suffix` is optional and can be used to ensure an order when applying
+patches from multiple files for the same component.
+- `+patchtype` is optional and can be used to pass one of the supported
+patch types. A missing value implies `+strategic`.
+- `.extension` is optional and can be used to facilitate user editing -
+e.g. `.yaml` would let the user edit the patch as a YAML document
+in their preferred editor.
+
+Examples:
+- `etcd.yaml`
+- `etcd+json`
+- `etcd+merge.json`
+- `etcd2.yaml`
+- `etcd2+merge.json`
+
+Alpha-numeric order will be used when applying patches from multiple
+files. This allows adjusting the order between patch types based on
+`suffix`.
+
+The format:
+- Must be documented briefly in `--help` screens.
+- Must be fully documented at the kubernetes/website (Beta graduation).
+- Would allow for basic usage (using the default `strategic` type),
+but also allow for advanced usage with patch ordering and the `json` type.
+- Would allow having arbitrary files that are ignored in the patch folder -
+e.g. having a `README.md`.
+
+### Support for multiple patches per file
+
+To support multiple patches per file, the proposal is to implement
```

Comment:
If we are going to support this I would drop the suffix part in the name, so there is only a way to provide multiple patches

Reply:
i prefer if we keep the optional suffix support because it allows changing the order in which different patch types are applied.

if we remove it the order will be fixed `json,merge,strategic`, which might now be ideal for everyone.

Reply:
Let's keep all options open for now.